File: TheSky/lua/op/AdventureView/chapterLayer.lua

chapterLayer = class("chapterLayer", function()
    local proxy = CCBProxy:create()
    local node  = CCBReaderLoad("ccbResources/ChapterView.ccbi",proxy, true,"ChapterViewOwner")
    local layer = tolua.cast(node,"CCLayer")
    return layer
end)

chapterLayer.bookId = nil
chapterLayer.chapterId = nil


ChapterViewOwner = ChapterViewOwner or {}
ccb["ChapterViewOwner"] = ChapterViewOwner

-- function chapterLayer.getChapterEnemiesCallback(url, rtnData)
--     getMainLayer():gotoChapterRob(rtnData["info"], chapterLayer.bookId, chapterLayer.chapterId)
-- end

-- function chapterLayer.chapterClick(tag)
--     local chapterPre = string.gsub(self.bookId, "book", "chapter")
--     local chapterId = string.format("%s_%02d", chapterPre, tag)
--     local count = 0
--     if chapterdata.chapters[self.bookId][chapterId] then
--         count = chapterdata.chapters[self.bookId][chapterId]
--     end
--     if count > 0 then
--         ShowText(HLNSLocalizedString("chapter.rob.haveOne"))
--     else
--         self.chapterId = chapterId
--         doActionFun("GET_CHAPTER_ENEMIES", {self.bookId, chapterId}, self.getChapterEnemiesCallback)
--     end
-- end
-- ChapterViewOwner["chapterClick"] = chapterLayer.chapterClick

-- function chapterLayer.combineItemClick()
--     local conf = skilldata:getSkillConfig(self.bookId)
--     local chapterNum = conf.chapternum
--     local flag = true
--     for i=1,chapterNum do
--         local chapterPre = string.gsub(self.bookId, "book", "chapter")
--         local chapterId = string.format("%s_%02d", chapterPre, i)
--         if not chapterdata.chapters[self.bookId][chapterId] or chapterdata.chapters[self.bookId][chapterId] < 1 then
--             flag = false
--             break
--         end
--     end
--     if not flag then
--         ShowText(HLNSLocalizedString("chapter.needBook"))
--     else
--         -- TODO 合成残章
--     end
-- end
-- ChapterViewOwner["combineItemClick"] = chapterLayer.combineItemClick

-- function chapterLayer:skillItemClick()
--     print(self.bookId)
--     getMainLayer():getParent():addChild(createChapterInfoLayer(self.bookId, -134))
-- end


function chapterLayer:refresh()

    local function getChapterEnemiesCallback(url, rtnData)
        getMainLayer():gotoChapterRob(rtnData["info"], self.bookId, self.chapterId)
    end

    local function chapterClick(tag)
        Global:instance():TDGAonEventAndEventData("adventure11")
        local chapterPre = string.gsub(self.bookId, "book", "chapter")
        local chapterId = string.format("%s_%02d", chapterPre, tag)

        local count = 0
        if chapterdata.chapters[self.bookId][chapterId] then
            count = chapterdata.chapters[self.bookId][chapterId]
        end
        if count > 0 then
            ShowText(HLNSLocalizedString("chapter.rob.haveOne"))
        else
            self.chapterId = chapterId
            doActionFun("GET_CHAPTER_ENEMIES", {self.bookId, chapterId}, getChapterEnemiesCallback)
        end
    end

    local function skillItemClick(tag, sender)
        getMainLayer():getParent():addChild(createChapterInfoLayer(self.bookId, -134))
    end

    local function combineSkillCallback(url,rtnData)
        runtimeCache.responseData = rtnData.info
        chapterdata:reduceCombineTime(self.bookId)
        getAdventureLayer():refreshAdventureLayer() 
        getMainLayer():gotoAdventure()
        local conf = skilldata:getSkillConfig(self.bookId)
        palyConvAnimationOnNode(self.bookId,conf.chapternum,true, getAdventureLayer())
    end

    local function getChaptersCallback(url, rtnData)
        if table.getTableCount(rtnData.info) then
            for k,v in pairs(rtnData.info) do
                chapterdata.chapters[k] = v
            end
        end
        getAdventureLayer():refreshAdventureLayer()
    end

    local function combineErrorCallback(url, rtnCode)
        if rtnCode == ErrorCodeTable.ERR_1106 then
            doActionFun("GET_BOOK_CHAPTER", {self.bookId}, getChaptersCallback)
        end
    end

    local function combineItemClick()
        Global:instance():TDGAonEventAndEventData("adventure10")
        local conf = skilldata:getSkillConfig(self.bookId)
        local flag = chapterdata:skillCanCombine(self.bookId)
        if not flag then
            ShowText(HLNSLocalizedString("chapter.needBook"))
        else
            doActionFun("COMBINE_CHAPTER", {self.bookId}, combineSkillCallback, combineErrorCallback)
        end
    end

    local combineItem = tolua.cast(ChapterViewOwner["combineItem"], "CCMenuItemImage")
    combineItem:registerScriptTapHandler(combineItemClick)

    local conf = skilldata:getSkillConfig(self.bookId)
    local skillName = tolua.cast(ChapterViewOwner["skillName"], "CCLabelTTF")
    skillName:setString(conf.name)

    local skillFrame = tolua.cast(ChapterViewOwner["skillFrame"], "CCMenuItemImage")
    skillFrame:setNormalSpriteFrame(CCSpriteFrameCache:sharedSpriteFrameCache():spriteFrameByName(string.format("frame_%d.png", conf.rank)))
    skillFrame:setSelectedSpriteFrame(CCSpriteFrameCache:sharedSpriteFrameCache():spriteFrameByName(string.format("frame_%d.png", conf.rank)))
    skillFrame:registerScriptTapHandler(skillItemClick)

    local skillIcon = tolua.cast(ChapterViewOwner["skillIcon"], "CCSprite")
    skillIcon:setVisible(false)
    local res = wareHouseData:getItemResource(self.bookId)
    if res.icon then
        local texture = CCTextureCache:sharedTextureCache():addImage(res.icon)
        if texture then
            skillIcon:setTexture(texture)
            skillIcon:setVisible(true)
        end
    end


    local chapterNum = conf.chapternum
    local layer = tolua.cast(ChapterViewOwner["layer_"..chapterNum], "CCLayer")
    layer:setVisible(true)
    local menu = tolua.cast(ChapterViewOwner["menu_"..chapterNum], "CCMenu")
    menu:setVisible(true)

    for i=1,chapterNum do
        local label = tolua.cast(ChapterViewOwner[string.format("%d_%d_count", chapterNum, i)], "CCLabelTTF")
        local count = 0
        local chapterPre = string.gsub(self.bookId, "book", "chapter")
        if chapterdata.chapters[self.bookId][string.format("%s_%02d", chapterPre, i)] then
            count = chapterdata.chapters[self.bookId][string.format("%s_%02d", chapterPre, i)]
        end
        label:setString(tostring(count))
        local item = tolua.cast(ChapterViewOwner[string.format("%d_%d_chapterItem", chapterNum, i)], "CCMenuItemImage")
        item:registerScriptTapHandler(chapterClick)
        -- renzhan
        if count == 0 then
            local cache = CCSpriteFrameCache:sharedSpriteFrameCache()
            cache:addSpriteFramesWithFile("ccbResources/treasureCard.plist")
            local light = CCSprite:createWithSpriteFrameName("treasureCard_roundFrame_1.png")
            local animFrames = CCArray:create()
            for j = 1, 3 do
                local frameName = string.format("treasureCard_roundFrame_%d.png",j)
                local frame = CCSpriteFrameCache:sharedSpriteFrameCache():spriteFrameByName(frameName)
                animFrames:addObject(frame)
            end
            local animation = CCAnimation:createWithSpriteFrames(animFrames, 0.2)
            local animate = CCAnimate:create(animation)
            light:runAction(CCRepeatForever:create(animate))
            item:addChild(light)
            light:setPosition(ccp(item:getContentSize().width / 2,item:getContentSize().height / 2))
        end
        -- 
    end

end

function createChapterLayer(bookId)
    local _chapterLayer = chapterLayer.new()
    _chapterLayer.bookId = bookId
    _chapterLayer:refresh()


    local function _onEnter()
    end

    local function _onExit()
    end

    --onEnter onExit
    local function _layerEventHandler(eventType)
        if eventType == "enter" then
            if _onEnter then _onEnter() end
        elseif eventType == "exit" then
            if _onExit then _onExit() end
        end
    end


    _chapterLayer:registerScriptHandler(_layerEventHandler)

    return _chapterLayer
end

File: TheSky/JavaScript/LZ/scenes/MainScene/Adventure/Chapter/chapterLayer.js
var chapterLayer = cc.Layer.extend({
	ctor:function(bookId){
		this._super();
		cc.BuilderReader.registerController("ChapterViewOwner", {});
		this.node = cc.BuilderReader.load(ccbi_res.ChapterView_ccbi, this);
		if (this.node != null) {
			this.addChild(this.node);
		}
		var size = this.node.getContentSize();
		this.setContentSize(size);
		this.bookId = bookId;
		this.chapterId;
		this.refresh();
	},
	refresh:function(){
		var cfg = SkillModule.getSkillConfig(this.bookId);
		this.skillName.setString(cfg.name);
		this.skillFrame.setNormalImage(new cc.Sprite("#frame_" + cfg.rank + ".png"));
		this.skillFrame.setSelectedImage(new cc.Sprite("#frame_" + cfg.rank + ".png"));
		this.rank_bg.setSpriteFrame(cc.spriteFrameCache.getSpriteFrame("rank_head_bg_" + cfg.rank + ".png"));
		this.skillIcon.visible = false;
		var res = common.getResource(cfg.icon);
		if (res.icon) {
			this.skillIcon.setTexture(res.icon);
			this.skillIcon.visible = true;
		} else {
			common.ShowText("找不到该图片")
		}
		var chapterNum = cfg.chapternum;
		this["layer_" + chapterNum].visible = true;
		this["menu_" + chapterNum].visible = true;
		this["menu_" + chapterNum].setEnabled(true);
		for (var i = 1; i < chapterNum + 1; i++) {
			var count = 0;
			var chapterPre = this.bookId.replace("book", "chapter");
			var fixI = common.fix(i, 2);
			if (chapterdata.chapters[this.bookId][chapterPre + "_" + fixI]) {
				this.chapterId = chapterPre + "_" + fixI;
				count = chapterdata.chapters[this.bookId][chapterPre + "_" + fixI];
			}
			this[chapterNum + "_" + i + "_count"].setString(count);
			var farme = this[chapterNum + "_" + i + "_frame"];
			this[chapterNum + "_" + i + "_rankbg"].setSpriteFrame(cc.spriteFrameCache.getSpriteFrame("rank_head_bg_" + cfg.rank + ".png"));
			var item = this[chapterNum + "_" + i + "_chapterItem"];
			item.setEnabled(true);
			if (count == 0) {
				var animFrames = [];
				var cache = cc.spriteFrameCache;
				cache.addSpriteFrames(common.formatLResPath("treasureCard.plist"));
				var light = new cc.Sprite("#treasureCard_roundFrame_1.png");
				for (var j = 1; j < 4; j++) {
					var frameName = "treasureCard_roundFrame_" + j + ".png";
					var frame = cc.spriteFrameCache.getSpriteFrame(frameName);
					animFrames.push(frame);
				}
				var animation = cc.Animation(animFrames, 0.2);
				var animate =new cc.Animate(animation);
				light.runAction(cc.repeatForever(animate));
				farme.addChild(light);
				light.setPosition(cc.p(farme.getContentSize().width / 2, farme.getContentSize().height / 2));
			}
		}
	},
	skillItemClick:function(){
		cc.director.getRunningScene().addChild(new SkillDetail(SkillModule.getSkillConfigInfo(this.bookId), 2));
	},
	getChapterEnemiesCallback:function(){
		postNotifcation(NOTI.GOTO_CHAPTER_ROB,{bookId:this.bookId,chapterId:this.chapterId})
	},
	chapterClick:function(sender){
		var tag = sender.getTag();
		var fixI = common.fix(tag, 2);
		var chapterPre = this.bookId.replace("book", "chapter");
		var chapterId = chapterPre + "_" + fixI;
		var count = 0;
		if (chapterdata.chapters[this.bookId][chapterId]) {
			count = chapterdata.chapters[this.bookId][chapterId];
		}
		if (count > 0) {
			common.showTipText(common.LocalizedString("chapter_rob_haveOne"));
		} else {
			ChapterModule.doOnGetFragEnemies(this.bookId, this.getChapterEnemiesCallback.bind(this));
		}
	},
	combineSkillCallback:function(){
		//var cfg = SkillModule.getSkillConfig(this.bookId);
		//TODO palyConvAnimationOnNode(self.bookId,conf.chapternum,true, getAdventureLayer())
		common.ShowText("合成成功");
		var page = AdventureModule.getPage(this.bookId);
		postNotifcation(this.bookId);
		postNotifcation(NOTI.GOTO_ADVENTURE, {page : page});
	},
	/**
	 * 残页合成
	 */
	combineItemClick:function(){
		var cfg = SkillModule.getSkillConfig(this.bookId);
		var flag = ChapterModule.skillCanCombine(this.bookId);
		if (!flag) {
			common.showTipText(common.LocalizedString("chapter_needBook"));
		} else {
			ChapterModule.doOnFragCombine(this.bookId, this.combineSkillCallback.bind(this), function(){
				common.ShowText("合成失败");
			});
		}
	},
	onEnter:function(){
		this._super();
		addObserver(this.bookId, "refresh", this);
	},
	onExit:function(){
		this._super();
		removeObserver(this.bookId, "refresh", this);
	},
})